src/database/catalog.cpp

```diff
@@ -1,69 +1,90 @@
 #include "catalog.h"
 
 namespace emerald {
-Catalog::Catalog() {}
-
-std::vector<Table*> Catalog::get_tables() {
-  return TableList;
+Catalog::Catalog() {
+  names = new std::unordered_map<int, std::string>();
+  db_fields = new std::unordered_map<int, DbFile*>();
+  p_fields = new std::unordered_map<int, std::string>();
 }
 
 void Catalog::AddTable(DbFile* file, std::string name, std::string pkeyField) {
-  Table* tab = new Table(file, name, pkeyField);
-  TableList.push_back(file);
-}
+  int table_id = file->get_id();
 
-int Catalog::get_table_id(std::string name) {
-  for (int i=0; i<TableList.size(); i++) {
-    if (TableList[i].name == name) {
-      return TableList[i].file->get_id();
+  for (auto const& elem: names) {
+    int id = elem.first;
+    if (names[id] == name) {
+      if (id != table_id) {
+        names.erase(id);
+        db_fields.erase(id);
+        p_fields.erase(id);
+      }
+      break;
     }
   }
-  // throw new NoSuchElementException();
+  names[table_id] = name;
+  db_fields[table_id] = file;
+  p_fields[table_id] = pkeyField;
 }
 
-std::string Catalog::get_table_name(int table_id) {
-  for (int i=0; i<TableList.size(); i++) {
-    if (TableList[i].file->get_id() == table_id) {
-      return TableList[i].name;
+void Catalog::AddTable(DbFile* file, std::string name) {
+  AddTable(file, name, "");
+}
+
+void Catalog::AddTable(DbFile* file) {
+  // addTable(file, (UUID.randomUUID()).toString());
+}
+
+int Catalog::get_table_id(std::string name) {
+  for (auto const& elem: names) {
+    int id = elem.first;
+    if (name[id] == name) {
+      return id;
     }
   }
-
   // throw new NoSuchElementException();
 }
-
 TupleDesc * Catalog::get_tuple_desc(int table_id) {
-  for (int i = 0; i < TableList.size(); i++) {
-    if (TableList[i].file->get_id() == table_id) {
-      return TableList[i].file->getTupleDesc();
-    }
+  if (db_fields.count(table_id)) {
+    return db_fields[table_id]->get_tuple_desc();
   }
   // throw new NoSuchElementException();
 }
 
-DbFile * get_db_file(int table_id) {
+std::string Catalog::get_table_name(int table_id) {
```

Comment:
Nice to see you created a map to store the mapping from table_id to name of the table too! Nice work!